### PR TITLE
Feature: add `report lesson` and `how to contribute` links in lesson footers

### PR DESCRIPTION
**Pull request by JustWaveThings**
## Because

Add links to lessons for learners to report issues with lessons instead of getting taken to make the changes themselves.

## This PR

- Adds a `Report an issue with this lesson` link next to the `Improve this lesson` link at the bottom of every lesson
- Adds a `Learn how to contribute` link as well.
- Adds a font awesome flag icon and a book icon for the respective links 
- Adds a small amount of padding between the links
- **Moves links above the completion component**
- **Adds responsiveness to show only icon with tooltip on mobile view**

## Additional Information

~This is a draft PR. More work needs to be done on where the report an issue will be linked to.~  ready for review.

discord link - https://discord.com/channels/505093832157691914/991035463949254767/1135930665502769226 

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behavior
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] I have verified all tests and linters pass after making these changes.
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If applicable, this PR includes new or updated automated tests

### CODE REVIEW

**Comment by JustWaveThings** [comment]
image
This is what it looks like locally.

**Comment by JustWaveThings** [comment]
I know earlier today I came across a 'report issue' form that was made back in 2022, with activity in the past month...  I think it initially was made by @ChargrilledChook if I'm not mistaken.  I can't seem to find it right this second, but that could be where the 'report issue' goes to.  Wiring all that up with ruby and whatnot is above my ability, however. 

Or, I could see about making an issue template on GitHub, much in the same way the `Improve this lesson on GitHub` does. That I feel might be in my ability. I like this option better than the form as it keeps the GH stuff in GH.

**Comment by ManonLef** [comment]
@JustWaveThings I'm not sure if this is what you're looking for, but if I clicked a "report an issue" button inside the curriculum, I'd probably expect to be taken here: https://github.com/TheOdinProject/curriculum/issues/new/choose

**Comment by JustWaveThings** [comment]
Yep, that's it! Right under my nose. Thanks @ManonLef!

**Comment by JustWaveThings** [comment]
Existing tests pass. No new tests written.

**Comment by JustWaveThings** [comment]
Please ignore #4058.

**Review by KevinMulhern — COMMENTED** [review]
Thanks for doing this @JustWaveThings.

3 links on this part of the page makes it a little too busy. Mobile shows it best:
Screenshot 2023-08-02 at 02 31 38

With all the links being related to the lesson content, they will work better being close to the content, that would also allow us to cut down on the link copy (just a mock up - needs more tweaking)
Screenshot 2023-08-02 at 02 55 20



We'll still have the issue with links being bunched up on mobile, so maybe we could get away with using just the icons and using tool tips to explain what they do:

Screenshot 2023-08-02 at 02 43 18

To add tool tips we'd need to add an aria label attribute `aria: { label: 'Learn how to contribute' }` and a hint CSS class to each of the links: `hint--top`.

Another, probably more future proof UI thing we could do is a pop-out menu. Github use it all over to reduce noise in their UI for this sort of stuff.  But I appreciate it might seem like overkill just for this.

**Comment by JustWaveThings** [comment]
Thanks for the feedback and quick turnaround @KevinMulhern!

I agree that since two of the three links are about the specific lesson, it does make sense to put them at the bottom of the lesson rather than between the completion component and the footer CTA.

I searched for 'pop out' menu ideas and didn't find anything that seemed appropriate for TOP's style. If you have a specific example of what you think would future-proof this, I'd be game to see what I could do. But if not, I don't think it's much of a compromise to go with the icons with tooltips for mobile and move the content above the completion component.

**Review by KevinMulhern — CHANGES_REQUESTED** [review]
Thanks for making those changes @JustWaveThings, the new background section doesn't look half bad but it's still a little too attention grabbing for me. 

Theres a lot already going on at the bottom of lessons, more so when it comes to project pages and submission lists. I'd be worried about overloading users with too much stuff that stands out. Especially when these links won't be used by a vast majority of our users. 

The links should be somewhere intuitive for when users go looking for them; but otherwise blend into the content as much as possible. 

Playing around with it: 
* Best placement for keeping it subtle seems to be right at the bottom with the same alignment as the lesson content.
* Since they are closer to the content and within that overall context - we can tighten up the copy by dropping the "this lesson" copy from the links. 
* I'd be very tempted to drop the "learn how to contribute" link entirely. I understand why it's included but it may be too much for every lesson. Out of all the links, it's probably the least important and we can expose users to it elsewhere. The report page has a link to the contributing guide and I believe one of our Git lessons has it as an assignment too.

Final result would be something like this:
Screenshot 2023-08-02 at 09 32 55

**Comment by JustWaveThings** [comment]
- I agree with dropping `how to contribute` and shortening `Report`
- Instead of `Edit on GitHub` what about `Improve on Github`

**Comment by JustWaveThings** [comment]
Hi.  I am going to see if there's anyone who can pair with me on this because while I got the desired layout for the XL view, the rest of the responsive tailwind is like whack-a-mole. I've never used Tailwind, but I've been using the docs and have had some success.

It appears that for the `lg` breakpoint, the article content is pushed to the middle with the `mx-auto` class, and whenever I try to emulate that with the div that holds the links, I get conflicts with the existing `xl` breakpoint.  

image
This is the `xl` breakpoint. Matches the expected.

For the `lg` breakpoint, if I try to stick to the grid with just some margin or padding, it seems to be inherited also by the `xl` breakpoint

image
this is the `lg` breakpoint

**Comment by KevinMulhern** [comment]
> I agree with dropping how to contribute and shortening Report
> Instead of Edit on GitHub what about Improve on Github

Great stuff! those copy changes look much nicer 🔥 

> Not sure if I have to @ you, but in case I should -- @KevinMulhern :)
No worries, Github don't make it very clear. As one of the reviewers I'll get notified automatically every-time something new happens on this PR, just to save you the trouble of doing it manually.

I'll have a think about how to handle the different breakpoints for tomorrow. Initial thought is extending the the content container component to have an optional footer where we can put these links.

**Comment by JustWaveThings** [comment]
Hey Kevin, are you around?

**Comment by JustWaveThings** [comment]
Wow.  That didn't turn out quite right...  
image

File: app/views/lessons/show.html.erb [hunk 1]
```diff
@@ -7,8 +7,29 @@
...
+          <%= link_to github_edit_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: {label: 'Improve this lesson on GitHub' }, class: 'hint--top underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
+ # Only display icon with tooltip at mobile width -%>
+          <span class="fab fa-github mr-1  "></span>
+ # Display text at larger viewport widths -%>
+          <span class=" ">
+            Improve on GitHub
+          </span>
+          <% end %>
+
+          <%= link_to github_report_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: { label: 'Report an issue with this lesson' }, class: '  underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5 ' do %>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:24]: Nit: Just a little bit of left over spacing we can remove

```suggestion
          <%= link_to github_report_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: { label: 'Report an issue with this lesson' }, class: 'underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
```

File: app/views/lessons/show.html.erb [hunk 2]
```diff
@@ -7,8 +7,29 @@
...
+        <%= render ContentContainerComponent.new(classes: 'xl:mx-0', data_attributes: { lesson_toc_target: 'lessonContent' }) do |component| %>
+        <%= @lesson.body.html_safe %>
+
+        <% component.with_footer do %>
+          <div class="py-10  xl:col-span-7 xl:col-start-2 lg:col-start-3 lg:col-span-7">
+          <%= link_to github_edit_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: {label: 'Improve this lesson on GitHub' }, class: 'hint--top underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
+ # Only display icon with tooltip at mobile width -%>
+          <span class="fab fa-github mr-1  "></span>
+ # Display text at larger viewport widths -%>
+          <span class=" ">
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:19]: Nit: Just a leftover empty class we can remove - we might be able to remove the span and just have the text on its own. But no harm in leaving it in either, totally up to you if you want to keep it.

```suggestion
          <span>
```

File: app/views/lessons/show.html.erb [hunk 3]
```diff
@@ -7,8 +7,29 @@
 
     <main class="grid grid-cols-12 gap-x-6" data-controller="lesson-toc" data-lesson-toc-item-classes-value="no-underline hover:text-gray-800 text-sm dark:hover:text-gray-300">
       <article class="col-span-full xl:col-span-7 xl:col-start-2">
-        <%= render ContentContainerComponent.new(classes: 'xl:mx-0', data_attributes: { lesson_toc_target: 'lessonContent' }) do %>
-          <%= @lesson.body.html_safe %>
+        <%= render ContentContainerComponent.new(classes: 'xl:mx-0', data_attributes: { lesson_toc_target: 'lessonContent' }) do |component| %>
+        <%= @lesson.body.html_safe %>
+
+        <% component.with_footer do %>
+          <div class="py-10  xl:col-span-7 xl:col-start-2 lg:col-start-3 lg:col-span-7">
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:14]: Nit: Now you have the container handling all the widths and break points, we can remove most of this 💪 
We can also change to only have padding top for some breathing room, the container is giving us padding bottom.

```suggestion
          <div class="pt-10">
```

File: app/views/lessons/show.html.erb [hunk 4]
```diff
@@ -7,8 +7,29 @@
...
+          <%= link_to github_edit_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: {label: 'Improve this lesson on GitHub' }, class: 'hint--top underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:15]: Do you want to keep the tool tips? I think we'd be ok without them. But I don't mind if you think they help and want to keep them.

```suggestion
          <%= link_to github_edit_url(@lesson), target: :_blank, rel: 'noreferrer noopener', class: 'underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
```

File: app/views/lessons/show.html.erb [hunk 5]
```diff
@@ -7,8 +7,29 @@
...
+          <span class=" ">
+            Improve on GitHub
+          </span>
+          <% end %>
+
+          <%= link_to github_report_url(@lesson), target: :_blank, rel: 'noreferrer noopener', aria: { label: 'Report an issue with this lesson' }, class: '  underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5 ' do %>
+
+          <span class="fas fa-flag mr-1 "></span>
+
+          <span class="  ">Report an issue</span>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:28]: Nit: Just a leftover empty class

```suggestion
          <span>Report an issue</span>
```

**Review by KevinMulhern — COMMENTED** [review]
heh looks like those comments are causing some issues. We don't need them now that we're showing the text on mobile so hopefully easy sorted 💪 

Just a few _very small_ final little tweaks from me. Thanks for all your work on this Fred - this is going to a really useful addition 🔥

**Comment by JustWaveThings** [comment]
So I don't know if this needs a separate PR, or if we can roll it into this, but after this https://discord.com/channels/505093832157691914/991035463949254767/1137026282429886494 comment today in chat, I think it might make sense to change `report issue` to 'report issue/suggestion'. 

Given that the link target mixes both reporting issues and making a suggestion:

image


I feel like we can make the link clearer by adopting `Report issue/suggestion` 

I am going to commit the change and if not, we can roll it back if we just want to keep the scope sighted on 'reporting issues'. If that is the case, I would want to rework the issue template to reflect that, and then create a suggestion template as well. 
Regardless, I do feel like the GH issue form could use some rework for this implementation regardless, but that would be a different PR to my understanding.

Thanks again for letting me help out with this feature! You have been so helpful and It truly has been a blast!

**Comment by Asartea** [comment]
I feel like while `report an issue` works, `report` starts sounding weird in the context of making a suggestion (`report a suggestion` suggests you are reporting the suggestion as inappropriate or something like that, especially when combined with the flag icon, which is also the icon we use for reporting submissions). What about `Report an issue or make a suggestion`?


Also while I appreciate being a bit late on this, I'm really not convinced that moving these links into the content area is the right choice UX wise. They now look like they are a (slightly misplaced) part of the main text, which they aren't.

A better place for them IMO would be to either center them or to center and move them into the same box as the lesson buttons, which makes the distinction between text and action much clearer

File: app/views/lessons/show.html.erb [hunk 6]
```diff
@@ -7,8 +7,29 @@
...
+          <%= link_to github_edit_url(@lesson), target: :_blank, rel: 'noreferrer noopener', class: 'underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
+
+          <span class='fab fa-github mr-1'></span>
+
+          <span>
+            Improve on GitHub
+          </span>
+          <% end %>
+
+          <%= link_to github_report_url(@lesson), target: :_blank, rel: 'noreferrer noopener', class: 'underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 px-5' do %>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:24]: Sorry I missed this on last pass through, just needs that left alignment we talked about on call.

```suggestion
          <%= link_to github_report_url(@lesson), target: :_blank, rel: 'noreferrer noopener', class: 'underline hover:no-underline text-gray-600 hover:text-gray-800 dark:text-gray-400 dark:hover:text-gray-300 pr-5' do %>
```

Last one, promise 😆

File: app/views/lessons/show.html.erb [hunk 7]
```diff
@@ -7,8 +7,29 @@
...
+
+          <span class='fas fa-flag mr-1'></span>
+
+          <span>Report an issue/suggestion</span>
```
Comment (KevinMulhern) [inline app/views/lessons/show.html.erb:28]: I absolutely get the reasoning behind adding suggestion and I don't mean to be difficult, it just reads a little awkward to me with report and suggestion together - get a second opinion on that lol.

In this case our issue template probably needs updated by dropping "suggestion" from it. Lately, we've been directing people to start a GitHub discussion when they have suggestions of things we could add or ideas of how to improve the curriculum. Our issues should only really be for stuff thats broken and immediately actionable.

One other thought 🤔 we could maybe link to the [choose screen](https://github.com/TheOdinProject/curriculum/issues/new/choose) instead to cover all our bases. And have "Suggest an improvement" as the text. That would pair really well with the "Improve on Github" link.

**Comment by JustWaveThings** [comment]
>I feel like while report an issue works, report starts sounding weird in the context of making a suggestion (report a suggestion suggests you are reporting the suggestion as inappropriate or something like that, especially when combined with the flag icon, which is also the icon we use for reporting submissions). What about Report an issue or make a suggestion?

I think Kevin's concern was making it too busy on smaller screen sizes. The PR started with 3 separate links.  Also, when you follow the link, it takes you to an issue form that reads to me like you could report a broken link or make a suggestion for the lesson, that's why I feel that some form of mentioning suggestion is appropriate for the link text.  I could be happy with any which way it's decided. 


> Also while I appreciate being a bit late on this, I'm really not convinced that moving these links into the content area is the right choice UX-wise. They now look like they are a (slightly misplaced) part of the main text, which they aren't.

I can see your point about them not being an organic part of the lesson text.  Would some sort of subtle background shading help separate them from the content and address that concern, do you think?  It seems we were pretty keen on moving the links above the completion component, so as to not further busy the busy footer / cta component. Perhaps we could cheat the padding to have the links closer to the lesson button component than the lesson material?  I think either/both of those would help create a visual break for that content that would signify: "this is related, but not part of the lesson content" Thoughts @Asartea ?

> A better place for them IMO would be to either center them or to center and move them into the same box as the lesson buttons, which makes the distinction between text and action much clearer

Personally, I don't like the idea of adding to that component. It would make it busier and less focused. As for alignment, I think the left alignment looks cleaner in this case.

**Review by KevinMulhern — COMMENTED** [review]
I can kind of see what @Asartea is saying about the links looking like they are part of the main text - One very simple thing we could do is change the font size to `text-sm`. That combined with the underline, different link color and the generous padding we've given this creates more than enough distinction imo. 

Screenshot 2023-08-04 at 17 12 25

File: app/views/lessons/show.html.erb (hunk 7 above)
Comment (JustWaveThings) [reply app/views/lessons/show.html.erb:28]: Yeah let's link to the choose screen. That covers the bases with report/suggest going to "Suggest an improvement"

**Comment by Asartea** [comment]
> I can kind of see what @Asartea is saying about the links looking like they are part of the main text - One very simple thing we could do is change the font size to text-sm. That combined with the underline, different link color and the generous padding we've given this creates more than enough distinction imo.

I still would _really_ like visual separation of some kind because it still looks somewhat out of place. Either some form of box surrounding it or more clearly aligning it out of the main text flow

**Comment by JustWaveThings** [comment]
This is what I have now, I don't know how the px-5 got back into the code. 
image

I think there is a lot of vertical separation (too much?) between the lesson, the links, and the complete lesson component. I think if add more to the links it will highlight them as more important

**Review by KevinMulhern — APPROVED** [review]
Padding looks great to me @JustWaveThings, it's consistent with the p-10 we have on the bottom of the container 👨‍🍳. Thanks for all your work on this, it will be appreciated by users and I personally appreciate something finally been done with that floating Github link 😆 

@Asartea while I disagree this time, I do appreciate your different point of view.  If you'd like, we can continue this in an issue or discussion.

**Pull request merged**